lib/services/data/budget/providers/project_budget_task_hierarchy_provider.dart

```diff
@@ -0,0 +1,142 @@
+import 'package:flutter_riverpod/flutter_riverpod.dart';
+import 'package:seren_ai_flutter/services/data/projects/providers/cur_selected_project_providers.dart';
+import 'package:seren_ai_flutter/services/data/tasks/providers/tasks_by_project_stream_provider.dart';
+import 'package:seren_ai_flutter/services/data/tasks/widgets/gantt/experimental/cur_project_tasks_hierarchy_provider.dart';
+
+final _curProjectBudgetTasksHierarchyProvider =
+    Provider<Map<String, TaskHierarchyInfo>>((ref) {
+  final projectId = ref.watch(curSelectedProjectIdNotifierProvider);
+  if (projectId == null) return {};
+
+  final tasks = ref.watch(tasksByProjectStreamProvider(projectId)).value ?? [];
```

Comment:
This is quite expensive- for any change to the tasks we have to recalculate all of this which appears to be a O(n^2) algorithm ... ideally anything expensive like this should be ref.read based so it doesn't keep on getting called so often.